### PR TITLE
Add documentation for running a packaged project

### PR DESCRIPTION
**Pull request by jmholzer**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

Resolves https://github.com/kedro-org/kedro/issues/1890

## Development notes

I have made two changes:

1. I have improved the instructions for running a packaged project in our tutorial.
2. I have added documentation for running a packaged project on Databricks (the original scope of #1890, though it is not explicitly mentioned in that ticket).

My additions under point 2. address all the problems raised in #1890. I feel that my additions under point 1. help to clarify the  documentation.

The changes made under point 2. are made at the end of the current documentation on deployment to Databricks. This file is being relocated in https://github.com/kedro-org/kedro/pull/2442. Its position will likely change in the future as part of https://github.com/kedro-org/kedro/issues/2443.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

### CODE REVIEW

**Comment by astrojuanlu** [comment]
```
/home/docs/checkouts/readthedocs.org/user_builds/kedro/envs/2425/lib/python3.7/site-packages/kedro/datasets/json/json_dataset.py:docstring of kedro.datasets.json.json_dataset.JSONDataSet:12: WARNING: Unexpected indentation.
/home/docs/checkouts/readthedocs.org/user_builds/kedro/envs/2425/lib/python3.7/site-packages/kedro/datasets/json/json_dataset.py:docstring of kedro.datasets.json.json_dataset.JSONDataSet:13: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/docs/checkouts/readthedocs.org/user_builds/kedro/envs/2425/lib/python3.7/site-packages/kedro/datasets/spark/spark_hive_dataset.py:docstring of kedro.datasets.spark.spark_hive_dataset.SparkHiveDataSet:9: WARNING: Bullet list ends without a blank line; unexpected unindent.
```

Looks like the problems come from docstrings that were not touched by this PR. Investigating.

**Comment by jmholzer** [comment]
Docs rendered at https://kedro--2425.org.readthedocs.build/en/2425/.

File: docs/source/deployment/databricks.md
Comment (stichbury) [inline docs/source/deployment/databricks.md:289]:
```suggestion
Use `configure_project()` to  run a packaged Kedro project interactively. Do **not** use `bootstrap_project()`, which is meant only for projects that haven't been packaged.
```

File: docs/source/deployment/databricks.md
Comment (stichbury) [inline docs/source/deployment/databricks.md:288]: I recommend that you remove this note and leave it as standard text, otherwise we'll have too many notes in the section

File: docs/source/deployment/databricks.md
Comment (stichbury) [inline docs/source/deployment/databricks.md:290]:
```suggestion

```

File: docs/source/deployment/databricks.md
Comment (stichbury) [inline docs/source/deployment/databricks.md:296]:
```suggestion

```

File: docs/source/deployment/databricks.md
Comment (stichbury) [inline docs/source/deployment/databricks.md:297]:
```suggestion
The entry-point of a packaged project is the `main()` method  of `src/<project_name>/__main__.py`, and it returns an exit code, which stops downstream processes from using its output. If you need to run your Kedro project interactively and use the results, consider using the `KedroSession` approach as shown above.
```

Comment (stichbury) [reply docs/source/deployment/databricks.md:297]: Can you add a link to the anchor point in the page where you should the KedroSession approach?

File: docs/source/tutorial/package_a_project.md
Comment (stichbury) [inline docs/source/tutorial/package_a_project.md:124]: Numbered bullets imply a sequence, that is, you have to do both. I would instead have an "either X or Y" statement as a sentence rather than a list.

File: docs/source/tutorial/package_a_project.md
Comment (stichbury) [inline docs/source/tutorial/package_a_project.md:141]: I think we are removing all mention of `.egg` from the docs, right, @astrojuanlu ?

Comment (stichbury) [inline docs/source/tutorial/package_a_project.md:141]:
```suggestion
To install the package from a `.whl` file, you need to have Python and `pip` on your machine, but do not need to have Kedro installed.
```

Comment (stichbury) [reply docs/source/tutorial/package_a_project.md:141]: I'd try to remove "recipient" as it's not an intuitive word.

File: docs/source/tutorial/package_a_project.md
Comment (stichbury) [inline docs/source/tutorial/package_a_project.md:144]:
```suggestion
Once the packaged project is installed, you will need to add:

* a `conf` folder
* a `data` folder if the pipeline loads/saves local data 
* a `logs` folder if the project uses logging.
```

**Review by stichbury — CHANGES_REQUESTED** [review]
Few changes suggested, sorry if it's a bit of a mess in there.

**Review by noklam — COMMENTED** [review]
It looks good in general, it would be great to check again to make sure the details are correct. I have left a few comments.

File: docs/source/deployment/databricks.md
Comment (jmholzer) [reply docs/source/deployment/databricks.md]: I modified the wording of this note in [9cd8ea6](https://github.com/kedro-org/kedro/pull/2425/commits/9cd8ea63dd4ba790585ed3dec9efc154bd383e2f) to make its message clearer to the user. In the new wording, I think it is clearer what is being referred to.

File: docs/source/tutorial/package_a_project.md
Comment (jmholzer) [reply docs/source/tutorial/package_a_project.md]: Thanks for explaining this, that makes sense.

File: docs/source/deployment/databricks.md
Comment (jmholzer) [reply docs/source/deployment/databricks.md:280]: It should be package name, I'm not sure how I let project_root get in here.

Comment (jmholzer) [reply docs/source/deployment/databricks.md]: Thanks for noticing, this was a weird mistake to make.

File: docs/source/tutorial/package_a_project.md
Comment (jmholzer) [reply docs/source/tutorial/package_a_project.md]: You're right that it doesn't expose `run`, it exposes the run API. I made this change in [30462c8](https://github.com/kedro-org/kedro/pull/2425/commits/30462c8f64c193ae89af58117f2b392a86a69008).

File: docs/source/tutorial/package_a_project.md
Comment (jmholzer) [reply docs/source/tutorial/package_a_project.md]: In my view, there are two things happening:

- The entry point to the script (which points to `package.__main__`) has the name `<package_name>`. It can be used to run the project.
- The entry point uses Kedro's `run` API, which is decorated using `click`. This allows options and arguments passed to the entry point to be consumed, for instance: `my-project --from-nodes A,B`. In my view, this qualifies as a CLI for the packaged project.

I feel that this 'CLI' approach to running a project is also more intuitive to beginner users than using `python`'s `-m` option. What do you think @stichbury?

> I think we should also explain what does python -m does, under the hood it picks up the package.__main__ which user actually has the ability to change before they package it.

I agree with you that this would be generally useful information, but I personally feel that this is too involved for our tutorial. A minority of advanced users may want to modify this behaviour and my view is that they are likely to have sufficient knowledge of Python packaging to achieve this independently. I would avoid overwhelming beginner and intermediate users with information.

File: docs/source/deployment/databricks.md
Comment (astrojuanlu) [inline docs/source/deployment/databricks.md:271]: Upcoming conflict with https://github.com/kedro-org/kedro/pull/2442

File: docs/source/tutorial/package_a_project.md
Comment (astrojuanlu) [inline docs/source/tutorial/package_a_project.md:124]: Worth adding an example snippet right below?

File: docs/source/tutorial/package_a_project.md
Comment (astrojuanlu) [inline docs/source/tutorial/package_a_project.md:146]:
```suggestion
* a `logs` folder if the project uses logging.

Alternatively, you can make use of the ``OmegaConfigLoader`` to run the configuration directly from the compressed .tar.gz configuration file by running
```

Fixing whitespace and replacing "they" by "you" (hope I understood this correctly)

File: docs/source/tutorial/package_a_project.md
Comment (astrojuanlu) [inline docs/source/tutorial/package_a_project.md:147]: To have code snippets inside fenced blocks, you can do

`````
````{note}
This is a note:

```
$ kedro run
```
````
`````

**Review by stichbury — APPROVED** [review]

File: docs/source/deployment/databricks.md
Comment (jmholzer) [reply docs/source/deployment/databricks.md:271]: You're right 🙂 I'll merge these in an optimal order once approved.

File: docs/source/tutorial/package_a_project.md
Comment (jmholzer) [reply docs/source/tutorial/package_a_project.md:124]: I thought about this, but it is difficult to justify as the commands are so short. Willing to be overruled but I would keep it as is.

File: docs/source/tutorial/package_a_project.md
Comment (jmholzer) [reply docs/source/tutorial/package_a_project.md:146]: Thanks for noticing!

File: docs/source/deployment/databricks.md
Comment (noklam) [inline docs/source/deployment/databricks.md:282]: This looks a bit weird, we already have `package_name` so I think we should pass in `package_name` instead of `<package_name>`, and the 2nd argument does not look right? Do we actually need that?

File: docs/source/tutorial/package_a_project.md
Comment (noklam) [inline docs/source/tutorial/package_a_project.md:124]: I am a bit confused about what it means. If we have a project name `spaceflights`, does it mean I should run `spaceflights` in the terminal? Does it actually works?

File: docs/source/tutorial/package_a_project.md
Comment (noklam) [inline docs/source/tutorial/package_a_project.md:127]: I am a bit hesitant about removing the mention of `python -m <package>` here.

1. We still have `python -m package` mentioned in a few place
2. This seems to be going to be the way how we deploy kedro, at the moment we have issues with a `click` so it doesn't work well on Databricks, but it's a special case. I personally

File: docs/source/tutorial/package_a_project.md
Comment (jmholzer) [reply docs/source/tutorial/package_a_project.md:124]: It works, in `setup.py` we define the default entry point for the script:

```python
entry_point = (
    "{{ cookiecutter.repo_name }} = {{ cookiecutter.python_package }}.__main__:main"
)
```

This entry point allows the user to run the package after installing it by calling the project's name (`cookiecutter.repo_name`).

File: docs/source/tutorial/package_a_project.md
Comment (jmholzer) [reply docs/source/tutorial/package_a_project.md:127]: Both this way and the way I describe are equivalent. They both call `__main__:main`. In fact, both expose the `click` options for `kedro run`. Try running `python -m <package_name> --help` and checking it for yourself.

I'm not too concerned about Databricks, as this PR also adds clear instructions for running a packaged project on that platform.

That being said, I've added this back in.

File: docs/source/tutorial/package_a_project.md
Comment (jmholzer) [reply docs/source/tutorial/package_a_project.md:124]: It might be confusing as I may have the wrong understanding of `package_name` and `project_name`.

Comment (noklam) [reply docs/source/tutorial/package_a_project.md:124]: Thanks for explaining, I wasn't awared that we register an entrypoint automatically.

Comment (jmholzer) [reply docs/source/tutorial/package_a_project.md:124]: I've looked into this more and decided that `<project_name>` is not right here, it would be inconsistent with how we use `<project_name>` elsewhere in the code base (we map it to `{{ cookiecutter.project_name }}`). I have changed this to `<repo_name>` and added a note to the user what is meant by `repo_name`.

I have modified the precedence of both ways of running a packaged project so that your preferred way (`python -m ...` is now first.

**Review by noklam — APPROVED** [review]
Thanks for making this PR and bear with my comments! I am happy about removing the undocumented entrypoint and let's keep it with the standard `python -m` way.

**Pull request merged**
